controlplane/pkg/model/client_connection.go

```diff
@@ -0,0 +1,164 @@
+package model
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/crossconnect"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/nsm"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/registry"
+	"sync"
+)
+
+type ClientConnectionState int8
+
+const (
+	ClientConnection_Ready      ClientConnectionState = 0
+	ClientConnection_Requesting ClientConnectionState = 1
+	ClientConnection_Healing    ClientConnectionState = 2
+	ClientConnection_Closing    ClientConnectionState = 3
+	ClientConnection_Closed     ClientConnectionState = 4
+)
+
+type ClientConnection struct {
+	ConnectionId    string
+	Xcon            *crossconnect.CrossConnect
+	RemoteNsm       *registry.NetworkServiceManager
+	Endpoint        *registry.NSERegistration
+	Dataplane       *Dataplane
+	ConnectionState ClientConnectionState
+	Request         nsm.NSMRequest
+	DataplaneState  DataplaneState
+}
+
+func (cc *ClientConnection) GetId() string {
+	if cc == nil {
+		return ""
+	}
+	return cc.ConnectionId
+}
+
+func (cc *ClientConnection) GetNetworkService() string {
+	if cc == nil {
+		return ""
+	}
+	return cc.Endpoint.GetNetworkService().GetName()
+}
+
+func (cc *ClientConnection) GetConnectionSource() nsm.NSMConnection {
+	if cc.Xcon.GetLocalSource() != nil {
+		return cc.Xcon.GetLocalSource()
+	} else {
+		return cc.Xcon.GetRemoteSource()
+	}
+}
+
+func (cc *ClientConnection) Clone() *ClientConnection {
+	if cc == nil {
+		return nil
+	}
+
+	var xcon *crossconnect.CrossConnect
+	if cc.Xcon != nil {
+		xcon = proto.Clone(cc.Xcon).(*crossconnect.CrossConnect)
+	}
+
+	var remoteNsm *registry.NetworkServiceManager
+	if cc.RemoteNsm != nil {
+		remoteNsm = proto.Clone(cc.RemoteNsm).(*registry.NetworkServiceManager)
+	}
+
+	var endpoint *registry.NSERegistration
+	if cc.Endpoint != nil {
+		endpoint = proto.Clone(cc.Endpoint).(*registry.NSERegistration)
+	}
+
+	var request nsm.NSMRequest
+	if cc.Request != nil {
+		request = cc.Request.Clone()
+	}
+
+	return &ClientConnection{
+		ConnectionId:    cc.ConnectionId,
+		Xcon:            xcon,
+		RemoteNsm:       remoteNsm,
+		Endpoint:        endpoint,
+		Dataplane:       cc.Dataplane.Clone(),
+		Request:         request,
+		ConnectionState: cc.ConnectionState,
+		DataplaneState:  cc.DataplaneState,
+	}
+}
+
+type clientConnectionDomain struct {
+	baseDomain
+	inner sync.Map
+}
+
+func (d *clientConnectionDomain) AddClientConnection(cc *ClientConnection) {
+	d.inner.Store(cc.ConnectionId, cc.Clone())
+	d.resourceAdded(cc.Clone())
```

Comment:
There can be a case when model is being modified concurrently from 2 different threads and so we can get the following order of execution:
```
T1: AddClientConnection(cc)                  | T2: DeleteClientConnection(cc.GetId())
-------------------------------------------------------------------------------------------
 d.inner.Store(cc.ConnectionId, cc.Clone())  | *
 *                                           | v := d.GetClientConnection(id)
 *                                           | if v == nil {
 *                                           |     return
 *                                           | }
 *                                           | d.inner.Delete(id)
 *                                           | d.resourceDeleted(v)
 d.resourceAdded(cc.Clone())                 | *
-------------------------------------------------------------------------------------------
```
And so all listeners will first receive `delete` event and then `add` event. I believe this can lead to some error state.

Reply:
Yes, you are right, I will fix it